# MODELLING

## Import the necessary libraries

In [1]:

# Data manipulation
import numpy as np
import pandas as pd

# File handling
from pathlib import Path

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Warnings
import warnings
warnings.filterwarnings("ignore")


# Scikit-learn


# Data splitting
from sklearn.model_selection import train_test_split

# Encoding
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
import re

# Scaling
from sklearn.preprocessing import StandardScaler

# Models
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

# Evaluation metrics
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    confusion_matrix,
    classification_report,
    roc_curve,
    precision_recall_curve
)

# Hyperparameter tuning
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV


# Gradient Boosting Models


from xgboost import XGBClassifier
from lightgbm import LGBMClassifier


# Class Imbalance

from imblearn.over_sampling import SMOTE


# Display options


pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

## 1. Load the data set after the feature engineering

In [2]:
# Define the output folder
OUTPUT_DIR = Path("./output")
train = pd.read_parquet(
    OUTPUT_DIR / "train_feature_engineered.parquet"
)

test = pd.read_parquet(
    OUTPUT_DIR / "test_feature_engineered.parquet"
)

print(train.shape)
print(test.shape)

(307511, 702)
(48744, 701)


## 2. Inspect the loaded data

In [3]:
# View dataset information
print("Train shape:", train.shape)
print("Test shape:", test.shape)

train.head()

Train shape: (307511, 702)
Test shape: (48744, 701)


,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,NAME_TYPE_SUITE,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,OWN_CAR_AGE,FLAG_MOBIL,FLAG_EMP_PHONE,FLAG_WORK_PHONE,FLAG_CONT_MOBILE,FLAG_PHONE,FLAG_EMAIL,OCCUPATION_TYPE,CNT_FAM_MEMBERS,REGION_RATING_CLIENT,REGION_RATING_CLIENT_W_CITY,WEEKDAY_APPR_PROCESS_START,HOUR_APPR_PROCESS_START,REG_REGION_NOT_LIVE_REGION,REG_REGION_NOT_WORK_REGION,LIVE_REGION_NOT_WORK_REGION,REG_CITY_NOT_LIVE_CITY,REG_CITY_NOT_WORK_CITY,LIVE_CITY_NOT_WORK_CITY,ORGANIZATION_TYPE,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,APARTMENTS_AVG,BASEMENTAREA_AVG,YEARS_BEGINEXPLUATATION_AVG,YEARS_BUILD_AVG,COMMONAREA_AVG,ELEVATORS_AVG,ENTRANCES_AVG,FLOORSMAX_AVG,FLOORSMIN_AVG,LANDAREA_AVG,LIVINGAPARTMENTS_AVG,LIVINGAREA_AVG,NONLIVINGAPARTMENTS_AVG,NONLIVINGAREA_AVG,APARTMENTS_MODE,BASEMENTAREA_MODE,YEARS_BEGINEXPLUATATION_MODE,YEARS_BUILD_MODE,COMMONAREA_MODE,ELEVATORS_MODE,ENTRANCES_MODE,FLOORSMAX_MODE,FLOORSMIN_MODE,LANDAREA_MODE,LIVINGAPARTMENTS_MODE,LIVINGAREA_MODE,NONLIVINGAPARTMENTS_MODE,NONLIVINGAREA_MODE,APARTMENTS_MEDI,BASEMENTAREA_MEDI,YEARS_BEGINEXPLUATATION_MEDI,YEARS_BUILD_MEDI,COMMONAREA_MEDI,ELEVATORS_MEDI,ENTRANCES_MEDI,FLOORSMAX_MEDI,FLOORSMIN_MEDI,LANDAREA_MEDI,LIVINGAPARTMENTS_MEDI,LIVINGAREA_MEDI,NONLIVINGAPARTMENTS_MEDI,NONLIVINGAREA_MEDI,FONDKAPREMONT_MODE,HOUSETYPE_MODE,TOTALAREA_MODE,WALLSMATERIAL_MODE,EMERGENCYSTATE_MODE,OBS_30_CNT_SOCIAL_CIRCLE,DEF_30_CNT_SOCIAL_CIRCLE,OBS_60_CNT_SOCIAL_CIRCLE,DEF_60_CNT_SOCIAL_CIRCLE,DAYS_LAST_PHONE_CHANGE,FLAG_DOCUMENT_2,FLAG_DOCUMENT_3,FLAG_DOCUMENT_4,FLAG_DOCUMENT_5,FLAG_DOCUMENT_6,FLAG_DOCUMENT_7,FLAG_DOCUMENT_8,FLAG_DOCUMENT_9,FLAG_DOCUMENT_10,FLAG_DOCUMENT_11,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR,BUREAU_DAYS_CREDIT_MEAN,BUREAU_DAYS_CREDIT_SUM,BUREAU_DAYS_CREDIT_MIN,BUREAU_DAYS_CREDIT_MAX,BUREAU_CREDIT_DAY_OVERDUE_MEAN,BUREAU_CREDIT_DAY_OVERDUE_SUM,BUREAU_CREDIT_DAY_OVERDUE_MIN,BUREAU_CREDIT_DAY_OVERDUE_MAX,BUREAU_DAYS_CREDIT_ENDDATE_MEAN,BUREAU_DAYS_CREDIT_ENDDATE_SUM,BUREAU_DAYS_CREDIT_ENDDATE_MIN,BUREAU_DAYS_CREDIT_ENDDATE_MAX,BUREAU_DAYS_ENDDATE_FACT_MEAN,BUREAU_DAYS_ENDDATE_FACT_SUM,BUREAU_DAYS_ENDDATE_FACT_MIN,BUREAU_DAYS_ENDDATE_FACT_MAX,BUREAU_AMT_CREDIT_MAX_OVERDUE_MEAN,BUREAU_AMT_CREDIT_MAX_OVERDUE_SUM,BUREAU_AMT_CREDIT_MAX_OVERDUE_MIN,BUREAU_AMT_CREDIT_MAX_OVERDUE_MAX,BUREAU_CNT_CREDIT_PROLONG_MEAN,BUREAU_CNT_CREDIT_PROLONG_SUM,BUREAU_CNT_CREDIT_PROLONG_MIN,BUREAU_CNT_CREDIT_PROLONG_MAX,BUREAU_AMT_CREDIT_SUM_MEAN,BUREAU_AMT_CREDIT_SUM_SUM,BUREAU_AMT_CREDIT_SUM_MIN,BUREAU_AMT_CREDIT_SUM_MAX,BUREAU_AMT_CREDIT_SUM_DEBT_MEAN,BUREAU_AMT_CREDIT_SUM_DEBT_SUM,BUREAU_AMT_CREDIT_SUM_DEBT_MIN,BUREAU_AMT_CREDIT_SUM_DEBT_MAX,BUREAU_AMT_CREDIT_SUM_LIMIT_MEAN,BUREAU_AMT_CREDIT_SUM_LIMIT_SUM,BUREAU_AMT_CREDIT_SUM_LIMIT_MIN,BUREAU_AMT_CREDIT_SUM_LIMIT_MAX,BUREAU_AMT_CREDIT_SUM_OVERDUE_MEAN,BUREAU_AMT_CREDIT_SUM_OVERDUE_SUM,BUREAU_AMT_CREDIT_SUM_OVERDUE_MIN,BUREAU_AMT_CREDIT_SUM_OVERDUE_MAX,BUREAU_DAYS_CREDIT_UPDATE_MEAN,BUREAU_DAYS_CREDIT_UPDATE_SUM,BUREAU_DAYS_CREDIT_UPDATE_MIN,BUREAU_DAYS_CREDIT_UPDATE_MAX,BUREAU_AMT_ANNUITY_MEAN,BUREAU_AMT_ANNUITY_SUM,BUREAU_AMT_ANNUITY_MIN,BUREAU_AMT_ANNUITY_MAX,BUREAU_BB_MONTHS_COUNT_MEAN,BUREAU_BB_MONTHS_COUNT_SUM,BUREAU_BB_MONTHS_COUNT_MIN,BUREAU_BB_MONTHS_COUNT_MAX,BUREAU_BB_MONTHS_MIN_MEAN,BUREAU_BB_MONTHS_MIN_SUM,BUREAU_BB_MONTHS_MIN_MIN,BUREAU_BB_MONTHS_MIN_MAX,BUREAU_BB_MONTHS_MAX_MEAN,BUREAU_BB_MONTHS_MAX_SUM,BUREAU_BB_MONTHS_MAX_MIN,BUREAU_BB_MONTHS_MAX_MAX,BUREAU_BB_STATUS_0_MEAN,BUREAU_BB_STATUS_0_SUM,BUREAU_BB_STATUS_0_MIN,BUREAU_BB_S

## 3. Check data types

In [4]:
# Check feature types
numeric_cols = train.select_dtypes(include="number").columns.tolist()

categorical_cols = train.select_dtypes(
    include=["object", "category", "string"]
).columns.tolist()

print(f"Numeric columns: {len(numeric_cols)}")
print(f"Categorical columns: {len(categorical_cols)}")

categorical_cols

Numeric columns: 686
Categorical columns: 16


['NAME_CONTRACT_TYPE',
 'CODE_GENDER',
 'FLAG_OWN_CAR',
 'FLAG_OWN_REALTY',
 'NAME_TYPE_SUITE',
 'NAME_INCOME_TYPE',
 'NAME_EDUCATION_TYPE',
 'NAME_FAMILY_STATUS',
 'NAME_HOUSING_TYPE',
 'OCCUPATION_TYPE',
 'WEEKDAY_APPR_PROCESS_START',
 'ORGANIZATION_TYPE',
 'FONDKAPREMONT_MODE',
 'HOUSETYPE_MODE',
 'WALLSMATERIAL_MODE',
 'EMERGENCYSTATE_MODE']

## 4. Separate the target and IDs

In [5]:
# Target variable
y = train["TARGET"]

# IDs
train_ids = train["SK_ID_CURR"]
test_ids = test["SK_ID_CURR"]

# Feature matrices
X = train.drop(columns=["TARGET", "SK_ID_CURR"])
X_test = test.drop(columns=["SK_ID_CURR"])

print(X.shape)
print(X_test.shape)

(307511, 700)
(48744, 700)


## 5. Identify categorical columns

In [6]:
categorical_cols = X.select_dtypes(
    include=["object", "category", "string"]
).columns.tolist()

print(f"Categorical columns: {len(categorical_cols)}")
categorical_cols

Categorical columns: 16


['NAME_CONTRACT_TYPE',
 'CODE_GENDER',
 'FLAG_OWN_CAR',
 'FLAG_OWN_REALTY',
 'NAME_TYPE_SUITE',
 'NAME_INCOME_TYPE',
 'NAME_EDUCATION_TYPE',
 'NAME_FAMILY_STATUS',
 'NAME_HOUSING_TYPE',
 'OCCUPATION_TYPE',
 'WEEKDAY_APPR_PROCESS_START',
 'ORGANIZATION_TYPE',
 'FONDKAPREMONT_MODE',
 'HOUSETYPE_MODE',
 'WALLSMATERIAL_MODE',
 'EMERGENCYSTATE_MODE']

## 6. Encoding

#### Step 1: Cardinality check

Check the number of unique categories of each categorical feature. This will help in identfying what type of encoding should I consider

In [7]:
# Check the number of unique values in each categorical feature

cardinality = (
    X[categorical_cols]
    .nunique()
    .sort_values(ascending=False)
    .to_frame(name="Unique Categories")
)

display(cardinality)

,Unique Categories
ORGANIZATION_TYPE,58
OCCUPATION_TYPE,19
NAME_INCOME_TYPE,8
NAME_TYPE_SUITE,8
WALLSMATERIAL_MODE,8
WEEKDAY_APPR_PROCESS_START,7
NAME_FAMILY_STATUS,6
NAME_HOUSING_TYPE,6
NAME_EDUCATION_TYPE,5
FONDKAPREMONT_MODE,5


### Cardinality Analysis

Cardinality analysis was performed to determine the number of unique categories in each categorical feature before choosing an encoding method. Most categorical variables had a small number of unique categories, ranging from 2 to 19. For example, `NAME_CONTRACT_TYPE`, `FLAG_OWN_CAR`, and `FLAG_OWN_REALTY` had only 2 unique categories, while `OCCUPATION_TYPE` had 19 categories.

The feature with the highest cardinality was `ORGANIZATION_TYPE`, which contained 58 unique categories. To reduce the number of dummy variables created during encoding and improve model efficiency, only the 15 most frequent organization types will be retained, while the remaining less frequent categories will be grouped into a single category labelled **"Other"**.

Overall, the dataset contains mostly low-cardinality categorical variables. Therefore, One-Hot Encoding is an appropriate encoding technique. Grouping the rare `ORGANIZATION_TYPE` categories before encoding helps reduce dimensionality while preserving the most important information in the data.

In [8]:
# Check categories for the two highest-cardinality features

for col in ["ORGANIZATION_TYPE"]:
    print("=" * 60)
    print(f"{col}")
    print("=" * 60)

    print(f"Number of unique categories: {X[col].nunique()}\n")

    display(
        X[col]
        .value_counts(dropna=False)
        .to_frame(name="Count")
    )

ORGANIZATION_TYPE
Number of unique categories: 58



,Count
ORGANIZATION_TYPE,
Business Entity Type 3,67992
XNA,55374
Self-employed,38412
Other,16683
Medicine,11193
Business Entity Type 2,10553
Government,10404
School,8893
Trade: type 7,7831


##### Step 2: Reduce cardinality of ORGANIZATION_TYPE

In [9]:
# Keep the top 15 categories
top_15_org = (
    X["ORGANIZATION_TYPE"]
    .value_counts()
    .nlargest(15)
    .index
)

# Group the rest into "Other"
for df in [X, X_test]:
    df["ORGANIZATION_TYPE"] = np.where(
        df["ORGANIZATION_TYPE"].isin(top_15_org),
        df["ORGANIZATION_TYPE"],
        "Other"
    )

print(X["ORGANIZATION_TYPE"].nunique())
print(X["ORGANIZATION_TYPE"].value_counts())

15
ORGANIZATION_TYPE
Business Entity Type 3    67992
Other                     65016
XNA                       55374
Self-employed             38412
Medicine                  11193
Business Entity Type 2    10553
Government                10404
School                     8893
Trade: type 7              7831
Kindergarten               6880
Construction               6721
Business Entity Type 1     5984
Transport: type 4          5398
Trade: type 3              3492
Industry: type 9           3368
Name: count, dtype: int64


In [10]:
print("Training set:")
print(X["ORGANIZATION_TYPE"].nunique())

print("\nTest set:")
print(X_test["ORGANIZATION_TYPE"].nunique())

Training set:
15

Test set:
15


## 7. Encode categorical features and scale numeric features

In [11]:
X = pd.get_dummies(
    X,
    columns=categorical_cols,
    drop_first=False,
    dtype="int8"
)

X_test = pd.get_dummies(
    X_test,
    columns=categorical_cols,
    drop_first=False,
    dtype="int8"
)

X, X_test = X.align(
    X_test,
    join="left",
    axis=1,
    fill_value=0
)

##### Aligning train and test

In [12]:
# Ensure train and test have identical columns

X, X_test = X.align(
    X_test,
    join="left",
    axis=1,
    fill_value=0
)

#### Verify the encoding

In [13]:
# Recompute feature types

numeric_cols = X.select_dtypes(include="number").columns.tolist()

categorical_cols = X.select_dtypes(
    include=["object", "category", "string"]
).columns.tolist()

print(f"Numeric columns: {len(numeric_cols)}")
print(f"Categorical columns: {len(categorical_cols)}")

Numeric columns: 787
Categorical columns: 0


#### Clean the column names

In [14]:


def clean_column_names(df):
    df = df.copy()

    df.columns = [
        re.sub(r'[^A-Za-z0-9_]', '_', col)
        for col in df.columns
    ]

    return df

In [15]:
X = clean_column_names(X)
X_test = clean_column_names(X_test)

## 8. Train / validation split

##### Splitting  the data

In [16]:


# Split into training and validation sets
X_train, X_valid, y_train, y_valid = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Training features:", X_train.shape)
print("Validation features:", X_valid.shape)
print("Training target:", y_train.shape)
print("Validation target:", y_valid.shape)

Training features: (246008, 787)
Validation features: (61503, 787)
Training target: (246008,)
Validation target: (61503,)


##### Verify the class distribution

This confirms that stratification worked correctly.

In [17]:
print("Original target distribution:")
print(y.value_counts(normalize=True))

print("\nTraining target distribution:")
print(y_train.value_counts(normalize=True))

print("\nValidation target distribution:")
print(y_valid.value_counts(normalize=True))

Original target distribution:
TARGET
0    0.919271
1    0.080729
Name: proportion, dtype: float64

Training target distribution:
TARGET
0    0.919271
1    0.080729
Name: proportion, dtype: float64

Validation target distribution:
TARGET
0    0.919272
1    0.080728
Name: proportion, dtype: float64


## 9. Handle class imbalance with SMOTE

In [18]:

# Handle Class Imbalance using SMOTE


smote = SMOTE(random_state=42)

X_train_smote, y_train_smote = smote.fit_resample(
    X_train,
    y_train
)

print("Before SMOTE")
print(y_train.value_counts())

print("\nAfter SMOTE")
print(y_train_smote.value_counts())

Before SMOTE
TARGET
0    226148
1     19860
Name: count, dtype: int64

After SMOTE
TARGET
0    226148
1    226148
Name: count, dtype: int64


In [20]:
print("Duplicate columns:", X.columns.duplicated().sum())

Duplicate columns: 0


To address the class imbalance in the training data, the Synthetic Minority Oversampling Technique (SMOTE) was applied after splitting the dataset into training and validation sets. This ensured that only the training data was balanced, preventing data leakage into the validation set.

## 10. Scale the Data (Logistic Regression Only)

In [21]:

# Feature Scaling

scaler = StandardScaler()

# Logistic Regression
X_train_scaled = scaler.fit_transform(X_train_smote)

X_valid_scaled = scaler.transform(X_valid)

X_test_scaled = scaler.transform(X_test) 

## 11. Create an Evaluation Function
To avoid writing the same evaluation code for every model.

In [22]:

# Model Evaluation Function


from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

results = []

def evaluate_model(model_name, y_true, y_pred, y_prob):

    accuracy = accuracy_score(y_true, y_pred)
    precision = precision_score(y_true, y_pred)
    recall = recall_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred)
    roc_auc = roc_auc_score(y_true, y_prob)

    results.append({
        "Model": model_name,
        "Accuracy": accuracy,
        "Precision": precision,
        "Recall": recall,
        "F1-score": f1,
        "ROC-AUC": roc_auc
    })

## 12. Train the Baseline Models


#### 1. Logistic Regression

In [23]:


log_model = LogisticRegression(
    random_state=42,
    max_iter=1000
)

log_model.fit(X_train_scaled, y_train_smote)

log_pred = log_model.predict(X_valid_scaled)

log_prob = log_model.predict_proba(X_valid_scaled)[:, 1]

evaluate_model(
    "Logistic Regression",
    y_valid,
    log_pred,
    log_prob
)

Logistic Regression was trained using the scaled SMOTE-balanced training data because it is sensitive to feature magnitudes.

#### 2. Decision Tree

In [24]:


dt_model = DecisionTreeClassifier(
    random_state=42
)

dt_model.fit(X_train_smote, y_train_smote)

dt_pred = dt_model.predict(X_valid)

dt_prob = dt_model.predict_proba(X_valid)[:,1]

evaluate_model(
    "Decision Tree",
    y_valid,
    dt_pred,
    dt_prob
)

#### 3. Random Forest

In [25]:


rf_model = RandomForestClassifier(
    random_state=42,
    n_jobs=-1
)

rf_model.fit(X_train_smote, y_train_smote)

rf_pred = rf_model.predict(X_valid)

rf_prob = rf_model.predict_proba(X_valid)[:,1]

evaluate_model(
    "Random Forest",
    y_valid,
    rf_pred,
    rf_prob
)

#### 4. XGBoost

In [26]:


xgb_model = XGBClassifier(
    random_state=42,
    eval_metric="logloss"
)

xgb_model.fit(X_train_smote, y_train_smote)

xgb_pred = xgb_model.predict(X_valid)

xgb_prob = xgb_model.predict_proba(X_valid)[:,1]

evaluate_model(
    "XGBoost",
    y_valid,
    xgb_pred,
    xgb_prob
)

#### 5. LightGBM

In [27]:


lgbm_model = LGBMClassifier(
    random_state=42
)

lgbm_model.fit(X_train_smote, y_train_smote)

lgbm_pred = lgbm_model.predict(X_valid)

lgbm_prob = lgbm_model.predict_proba(X_valid)[:,1]

evaluate_model(
    "LightGBM",
    y_valid,
    lgbm_pred,
    lgbm_prob
)

[LightGBM] [Info] Number of positive: 226148, number of negative: 226148
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 1.716815 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 133760
[LightGBM] [Info] Number of data points in the train set: 452296, number of used features: 772
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000


Decision Tree, Random Forest, XGBoost, and LightGBM were trained using the SMOTE-balanced training data without feature scaling, as tree-based algorithms are generally not affected by differences in feature scale.

#### Compare the Models

In [28]:
results_df = (
    pd.DataFrame(results)
    .sort_values("ROC-AUC", ascending=False)
    .reset_index(drop=True)
)

results_df

,Model,Accuracy,Precision,Recall,F1-score,ROC-AUC
0,LightGBM,0.919695,0.565000,0.022759,0.043756,0.779973
1,Logistic Regression,0.919760,0.546296,0.035650,0.066931,0.776599
2,XGBoost,0.918801,0.477165,0.061027,0.108214,0.773136
3,Random Forest,0.919288,0.514286,0.003625,0.007200,0.735794
4,Decision Tree,0.844739,0.142879,0.184693,0.161117,0.543698


The baseline model comparison shows that LightGBM achieved the best overall performance with a ROC-AUC score of 0.780, making it the strongest model for predicting loan default. Logistic Regression followed closely with a ROC-AUC of 0.777, indicating that the engineered features were effective enough for a simple linear model to perform competitively. XGBoost also performed well with a ROC-AUC of 0.773, while Random Forest achieved a lower ROC-AUC of 0.736. The Decision Tree was the weakest model, with a ROC-AUC of 0.544, which is only slightly better than random guessing.

Although the top three models achieved high accuracy (around 92%), they recorded relatively low recall values. This indicates that the models correctly classified most non-default customers but identified only a small proportion of customers who actually defaulted. Since credit risk prediction is an imbalanced classification problem, ROC-AUC provides a more reliable measure of model performance than accuracy alone. Based on these baseline results, LightGBM was selected as the best-performing model and will be further improved through hyperparameter tuning.

## 13. Hyperparameter tuning

Hyperparameter tuning is important because the default settings are not necessarily the best settings for our dataset. By testing different combinations, we can improve the model's ability to distinguish between customers who are likely to default and those who are not.

In our project, the baseline LightGBM achieved a ROC-AUC of 0.780. Hyperparameter tuning gives the opportunity to improve this performance while also controlling the model's complexity and reducing the risk of overfitting.

In [ ]:


# Create the LightGBM model
lgbm_model = LGBMClassifier(
    objective="binary",
    random_state=42,
    n_jobs=1,
    verbosity=-1
)

# Define a smaller hyperparameter search space
param_grid = {
    "n_estimators": [100, 200],
    "learning_rate": [0.03, 0.05, 0.1],
    "num_leaves": [15, 31],
    "max_depth": [5, 10],
    "min_child_samples": [50, 100],
    "subsample": [0.8, 1.0],
    "colsample_bytree": [0.8, 1.0]
}

# Randomized search with fewer folds and fewer combinations
lgbm_search = RandomizedSearchCV(
    estimator=lgbm_model,
    param_distributions=param_grid,
    n_iter=3,
    scoring="roc_auc",
    cv=2,
    random_state=42,
    n_jobs=1,
    verbose=1
)

# Fit the search
lgbm_search.fit(
    X_train_smote,
    y_train_smote
)

# Display the best parameters
print("\nBest Parameters:")
print(lgbm_search.best_params_)

# Display the best cross-validation score
print("\nBest Cross-Validation ROC-AUC:")
print(round(lgbm_search.best_score_, 4))

# Get the best model
best_lgbm = lgbm_search.best_estimator_

# Evaluate on the untouched validation set
lgbm_tuned_pred = best_lgbm.predict(X_valid)

lgbm_tuned_prob = best_lgbm.predict_proba(
    X_valid
)[:, 1]

evaluate_model(
    "Tuned LightGBM",
    y_valid,
    lgbm_tuned_pred,
    lgbm_tuned_prob
)

# Display model comparison
results_df = (
    pd.DataFrame(results)
    .sort_values("ROC-AUC", ascending=False)
    .reset_index(drop=True)
)

display(results_df)

Fitting 2 folds for each of 3 candidates, totalling 6 fits
